src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -15,26 +19,74 @@ public class ParamsSemanticModel
+        public List<IDiagnostic> allDiagnostics { get; } = new();
+        public IFileResolver fileResolver { get; }
         
-        public ParamsSemanticModel(BicepParamFile bicepParamFile, Compilation? bicepCompilation = null)
+        public ParamsSemanticModel(BicepParamFile bicepParamFile, IFileResolver fileResolver, Func<Uri, Compilation>? getCompilation = null)
         {
+            this.fileResolver = fileResolver;
+
+            //parse using statement and link bicep template 
+            this.BicepParamFile = bicepParamFile;
+
+            Uri? bicepFileUri = TryGetBicepFileUri();
+            if(bicepFileUri is {} && getCompilation is {})
+            {   
+                this.BicepCompilation = getCompilation(bicepFileUri);
+            }
+
+            //binder logic
             var paramsSymbolContext = new ParamsSymbolContext(this);
             ParamsSymbolContext = paramsSymbolContext;
             this.BicepParamFile = bicepParamFile;
-            this.BicepCompilation = bicepCompilation;
             this.ParamBinder = new(bicepParamFile, paramsSymbolContext); 
             this.ParamsTypeManager = new(this, ParamBinder);
             // name binding is done
             // allow type queries now
             paramsSymbolContext.Unlock();
         }
 
-        public IEnumerable<IDiagnostic> GetDiagnostics()
-            => BicepParamFile.ProgramSyntax.GetParseDiagnostics();
+        public IEnumerable<IDiagnostic> GetAllDiagnostics()
+            => this.allDiagnostics.Concat(BicepParamFile.ProgramSyntax.GetParseDiagnostics());
         
         /// <summary>
         /// Gets the file that was compiled.
         /// </summary>
         public ParamFileSymbol Root => this.ParamBinder.ParamFileSymbol;
+
+        private Uri? TryGetBicepFileUri()
+        {
+            //TODO: consolidate this into one method
+            var usingDeclarations = BicepParamFile.ProgramSyntax.Children.OfType<UsingDeclarationSyntax>();
+
+            if(usingDeclarations.Count() == 0)
+            {
+                this.allDiagnostics.Add(new DiagnosticBuilder.DiagnosticBuilderInternal(new TextSpan(0, 0)).UsingDeclarationNotSpecified());
+                return null;
+            }
+            
+            if(usingDeclarations.Count() > 1)
+            {
+                this.allDiagnostics.Add(new DiagnosticBuilder.DiagnosticBuilderInternal(new TextSpan(0, 0)).MoreThenOneUsingDeclarationSpecified());
+                return null;
+            }
+
+            if(!PathHelper.TryGetUsingPath(usingDeclarations.FirstOrDefault()!, out var bicepFilePath, out var failureBuilder))
+            {       
+                var diagnostic = failureBuilder(new DiagnosticBuilder.DiagnosticBuilderInternal(new TextSpan(0, 0)));
+                this.allDiagnostics.Add(diagnostic);
+                return null;
+            }
+
+            Uri.TryCreate(BicepParamFile.FileUri, bicepFilePath, out var bicepFileUri);
```

Comment:
Please check the return value of `Uri.TryCreate` and raise a diagnostic if it fails. This should also allow you to avoid the `bicepFileUri!` nullable -> non-null conversion on line 83.

You could always combine these if needed:
```csharp
if (!Uri.TryCreate(BicepParamFile.FileUri, bicepFilePath, out var bicepFileUri) ||
    !fileResolver.FileExists(bicepFileUri))
{
    ...
}
```